

*   Install library sklearn, nltk, dan sastrawi





In [492]:
!pip install sklearn
!pip install nltk
!pip install sastrawi

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/




*   Import library pandas, numpy, nltk, dan re





In [493]:
import pandas as pd
import numpy as np
import nltk
from nltk.corpus import stopwords
import re



*   Load file dataset dari github menggungakan fungsi dari library pandas



In [494]:
df = pd.read_csv("https://raw.githubusercontent.com/ThoriqFathu/prosaindata/main/dataset-baru.csv")
df

Unnamed: 0                                            Abstrak   Kategori
0             0  Sistem  informasi  akademik  (SIAKAD) merupaka...        RPL
1             1  Berjalannya koneksi jaringan komputer dengan l...        RPL
2             2  Web server adalah sebuah perangkat lunak serve...        RPL
3             3  Penjadwalan  kuliah  di  Perguruan  Tinggi  me...  KOMPUTASI
4             4  Seiring perkembangan teknologi yang ada diduni...        RPL
..          ...                                                ...        ...
800         804  Investasi saham selama ini memiliki resiko ker...  KOMPUTASI
801         805  Information Retrieval (IR) merupakan pengambil...  KOMPUTASI
802         806  Klasifikasi citra merupakan proses pengelompok...  KOMPUTASI
803         807  Identifikasi atribut pejalan kaki merupakan sa...  KOMPUTASI
804         808  Topik deteksi objek telah menarik perhatian ya...  KOMPUTASI

[805 rows x 3 columns]



*   Memisahkan kolom target (kategori) dengan kolom abstrak yang nantinya akan dijadikan feature/attribut



In [495]:
text = df['Abstrak']
Y = df['Kategori'].values
text

0      Sistem  informasi  akademik  (SIAKAD) merupaka...
1      Berjalannya koneksi jaringan komputer dengan l...
2      Web server adalah sebuah perangkat lunak serve...
3      Penjadwalan  kuliah  di  Perguruan  Tinggi  me...
4      Seiring perkembangan teknologi yang ada diduni...
                             ...                        
800    Investasi saham selama ini memiliki resiko ker...
801    Information Retrieval (IR) merupakan pengambil...
802    Klasifikasi citra merupakan proses pengelompok...
803    Identifikasi atribut pejalan kaki merupakan sa...
804    Topik deteksi objek telah menarik perhatian ya...
Name: Abstrak, Length: 805, dtype: object



*   Membuat fungsi untuk merubah seluruh huruf menjadi huruf kecil



In [496]:
def case_folding(comment):
    comment = comment.lower()
    return comment
text = text.apply(case_folding)
text

0      sistem  informasi  akademik  (siakad) merupaka...
1      berjalannya koneksi jaringan komputer dengan l...
2      web server adalah sebuah perangkat lunak serve...
3      penjadwalan  kuliah  di  perguruan  tinggi  me...
4      seiring perkembangan teknologi yang ada diduni...
                             ...                        
800    investasi saham selama ini memiliki resiko ker...
801    information retrieval (ir) merupakan pengambil...
802    klasifikasi citra merupakan proses pengelompok...
803    identifikasi atribut pejalan kaki merupakan sa...
804    topik deteksi objek telah menarik perhatian ya...
Name: Abstrak, Length: 805, dtype: object



*   Membuat fungsi untuk menghilangkan link/url



In [497]:
def linkNormalize(text):
    text = re.sub(r"\s—\s", "", text)
    text = re.sub(r"http\S+", "", text)
    return text
text = text.apply(linkNormalize)
text

0      sistem  informasi  akademik  (siakad) merupaka...
1      berjalannya koneksi jaringan komputer dengan l...
2      web server adalah sebuah perangkat lunak serve...
3      penjadwalan  kuliah  di  perguruan  tinggi  me...
4      seiring perkembangan teknologi yang ada diduni...
                             ...                        
800    investasi saham selama ini memiliki resiko ker...
801    information retrieval (ir) merupakan pengambil...
802    klasifikasi citra merupakan proses pengelompok...
803    identifikasi atribut pejalan kaki merupakan sa...
804    topik deteksi objek telah menarik perhatian ya...
Name: Abstrak, Length: 805, dtype: object



*   Membuat fungsi untuk menghilangkan tanda baca



In [498]:
symbols = ",!\"#$%&()*+-.…/:;<=>?@[\]^_`{|}~\n"
def cleaning_text(comment):
    for i in symbols:
        comment = np.char.replace(comment, i, ' ')
    return comment
text = text.apply(cleaning_text)
text

0      sistem  informasi  akademik   siakad  merupaka...
1      berjalannya koneksi jaringan komputer dengan l...
2      web server adalah sebuah perangkat lunak serve...
3      penjadwalan  kuliah  di  perguruan  tinggi  me...
4      seiring perkembangan teknologi yang ada diduni...
                             ...                        
800    investasi saham selama ini memiliki resiko ker...
801    information retrieval  ir  merupakan pengambil...
802    klasifikasi citra merupakan proses pengelompok...
803    identifikasi atribut pejalan kaki merupakan sa...
804    topik deteksi objek telah menarik perhatian ya...
Name: Abstrak, Length: 805, dtype: object



*   Membuat fungsi untuk menghilangkan kata yang tidak penting menggunakan library nltk



In [499]:
nltk.download('stopwords')
list_stopwords = stopwords.words('indonesian')
def stopword_removal(comment):
    comment = comment.split()
    data = [word for word in comment if word not in list_stopwords]
    kalimat = ""
    for i in data:
      kalimat += i
      kalimat += " "
    return kalimat


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [500]:
text = text.apply(stopword_removal)

0      sistem informasi akademik siakad sistem inform...
1      berjalannya koneksi jaringan komputer lancar g...
2      web server perangkat lunak server berfungsi me...
3      penjadwalan kuliah perguruan kompleks permasal...
4      seiring perkembangan teknologi didunia muncul ...
                             ...                        
800    investasi saham memiliki resiko kerugian perge...
801    information retrieval ir pengambilan informasi...
802    klasifikasi citra proses pengelompokan piksel ...
803    identifikasi atribut pejalan kaki salah peneli...
804    topik deteksi objek menarik perhatian perkemba...
Name: abstrak, Length: 805, dtype: object



*   Membuat fungsi untuk menjadikan kata ke kata dasar dengan library sastrawi



In [501]:
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

factory = StemmerFactory()
stemmer = factory.create_stemmer()
def stemmingNormalize(comment):
    comment = comment.split()
    data = []
    for term in comment:
        term = stemmer.stem(term)
        data.append(term)
    kalimat = ""
    for i in data:
      kalimat += i
      kalimat += " "
    return kalimat
text = text.apply(stemmingNormalize)
text

0      sistem informasi akademik siakad sistem inform...
1      jalan koneksi jaring komputer lancar ganggu ha...
2      web server perangkat lunak server fungsi terim...
3      jadwal kuliah guru kompleks masalah variabel t...
4      iring kembang teknologi dunia muncul teknologi...
                             ...                        
800    investasi saham milik resiko rugi gera harga s...
801    information retrieval ir ambil informasi simpa...
802    klasifikasi citra proses kelompok piksel citra...
803    identifikasi atribut pejal kaki salah teliti k...
804    topik deteksi objek tarik perhati kembang tekn...
Name: Abstrak, Length: 805, dtype: object



*   Melakukan ekstraksi fitur menggunakan fungsi TF-IDF dari library sklearn
*   Hasil ekstraksi fitur dijadikan dataframe menggunakan fungsi dari library pandas





In [521]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()
tfidf_separate = tfidf_vectorizer.fit_transform(text)

df_tfidf = pd.DataFrame(
    tfidf_separate.toarray(), columns=tfidf_vectorizer.get_feature_names_out(), index=df.index
)
X = df_tfidf.values
df_tfidf

00  000  00001  0001  000505341  000506436  000792353  00086  00089  \
0    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
1    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
2    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
3    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
4    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
..   ...  ...    ...   ...        ...        ...        ...    ...    ...   
800  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
801  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
802  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
803  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
804  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   

     001  ...  zara  zat  zcz   zf  zona  zone  zoning  zoom  zucara  zungu  
0    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
1    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
2    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
3    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
4    0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
..   ...  ...   ...  ...  ...  ...   ...   ...     ...   ...     ...    ...  
800  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
801  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
802  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
803  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  
804  0.0  ...   0.0  0.0  0.0  0.0   0.0   0.0     0.0   0.0     0.0    0.0  

[805 rows x 6570 columns]



*   Melakukan spilt dataset menjadi data training sebesar 80% dan data test sebesar 20% menggunkan library sklearn



In [503]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( 
    X, Y, test_size = 0.2, random_state = 100)
print("Jumlah Data training : ", len(X_train))
print("Jumlah Data test : ", len(X_test))

Jumlah Data training :  644
Jumlah Data test :  161




**Melakukan skenario uji coba dengan mereduksi fitur dengan n component sebanyak 2,4,6,8,10**

*   n component = 2





In [504]:
from sklearn.decomposition import PCA

pca = PCA(n_components =2)

X_train_2 = pca.fit_transform(X_train)
X_test_2 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.02001121, 0.0140943 ])

In [505]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_2, y_train)

KNeighborsClassifier()

In [506]:
from sklearn.metrics import accuracy_score
y_pred = classifier.predict(X_test_2)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  75.15527950310559




*   n component = 4



In [507]:
pca = PCA(n_components =4)

X_train_4 = pca.fit_transform(X_train)
X_test_4 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.02001121, 0.01409429, 0.01303834, 0.01061409])

In [508]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_4, y_train)

KNeighborsClassifier()

In [509]:
y_pred = classifier.predict(X_test_4)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  72.04968944099379




*   n component = 6



In [510]:
pca = PCA(n_components = 6)
X_train_6 = pca.fit_transform(X_train)
X_test_6 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.02001121, 0.01409431, 0.01303847, 0.01061483, 0.00965392,
       0.00858264])

In [511]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_6, y_train)

KNeighborsClassifier()

In [512]:
y_pred = classifier.predict(X_test_6)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  76.3975155279503




*   n component = 8



In [513]:
pca = PCA(n_components = 8)

X_train_8 = pca.fit_transform(X_train)
X_test_8 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.02001121, 0.01409432, 0.0130385 , 0.01061461, 0.00965414,
       0.00858613, 0.00807648, 0.00733245])

In [514]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_8, y_train)

KNeighborsClassifier()

In [515]:
y_pred = classifier.predict(X_test_8)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  81.36645962732919




*   n component = 10



In [516]:
pca = PCA(n_components = 10)

X_train_10 = pca.fit_transform(X_train)
X_test_10 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.02001121, 0.01409431, 0.01303847, 0.01061521, 0.00965423,
       0.00857989, 0.00808247, 0.00734795, 0.00705504, 0.00678476])

In [517]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_10, y_train)

KNeighborsClassifier()

In [518]:
y_pred = classifier.predict(X_test_10)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  83.85093167701864




*   Tanpa reduksi



In [519]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [520]:
y_pred = classifier.predict(X_test)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  83.22981366459628


**KESIMPULAN**

*   Berdasarkan hasil skenario uji coba dapat disimpulkan bahwasanya dengan melakukan reduksi dimensi n componen sebanyak 8 dapat menghasilkan akurasi yang tidak jauh berbeda dengan n componen sebanyak 6570 dengan nilai akurasi diatas 80% menggunakan model KNN dengan nilai k sebesar 5
*   Dengan demikian metode PCA dapat diterapkan untuk mempercepat proses komputasi



